tasks/human_eval_005.rs

```diff
@@ -5,11 +5,119 @@ HumanEval/5
 /*
 ### VERUS BEGIN
 */
+use vstd::assert_seqs_equal;
 use vstd::prelude::*;
 
 verus! {
 
-// TODO: Put your solution (the specification, implementation, and proof) to the task here
+/// Specification for inserting a number 'delimeter' between every two consecutive elements
+/// of input sequence `numbers'
+pub open spec fn intersperse_spec(numbers: Seq<u64>, delimeter: u64) -> Seq<u64>
+    decreases numbers.len(),
+{
+    if numbers.len() <= 1 {
+        numbers
+    } else {
+        intersperse_spec(numbers.drop_last(), delimeter) + seq![delimeter, numbers.last()]
+    }
+}
+
+// We use these two functions to provide valid triggers for the quantifiers in intersperse_quantified
+spec fn even(i: int) -> int {
+    2 * i
+}
+
+spec fn odd(i: int) -> int {
+    2 * i + 1
+}
+
+// This quantified formulation of intersperse is easier to reason about in the implementation's loop
+spec fn intersperse_quantified(numbers: Seq<u64>, delimeter: u64, interspersed: Seq<u64>) -> bool {
+    (if numbers.len() == 0 {
+        interspersed.len() == 0
+    } else {
+        interspersed.len() == 2 * numbers.len() - 1
+    }) && (forall|i: int| 0 <= i < numbers.len() ==> #[trigger] interspersed[even(i)] == numbers[i])
+        && (forall|i: int|
+        0 <= i < numbers.len() - 1 ==> #[trigger] interspersed[odd(i)] == delimeter)
+}
+
+proof fn intersperse_spec_len(numbers: Seq<u64>, delimeter: u64)
+    ensures
+        numbers.len() > 0 ==> intersperse_spec(numbers, delimeter).len() == 2 * numbers.len() - 1,
+    decreases numbers.len(),
+{
+    if numbers.len() > 0 {
+        intersperse_spec_len(numbers.drop_last(), delimeter);
+    }
+}
+
+// Show that the two descriptions of intersperse are equivalent
+proof fn intersperse_quantified_is_spec(numbers: Seq<u64>, delimeter: u64, interspersed: Seq<u64>)
+    requires
+        intersperse_quantified(numbers, delimeter, interspersed),
+    ensures
+        interspersed =~= intersperse_spec(numbers, delimeter),
+    decreases numbers.len(),
+{
+    let is = intersperse_spec(numbers, delimeter);
+    if numbers.len() == 0 {
+    } else if numbers.len() == 1 {
+        assert(interspersed.len() == 1);
+        assert(interspersed[even(0)] == numbers[0]);
+    } else {
+        intersperse_quantified_is_spec(
+            numbers.drop_last(),
+            delimeter,
+            interspersed.take(interspersed.len() - 2),
+        );
+        intersperse_spec_len(numbers, delimeter);
+        assert_seqs_equal!(is == interspersed, i => {
```

Comment:
I think `assert_seqs_equal!` is deprecated now, and we're generally supposed to use `=~=` instead.

Reply:
Interesting. I hadn't realized `assert_seqs_equal!` was so general.